core/src/main/scala/org/allenai/common/ParIterator.scala

```diff
@@ -0,0 +1,58 @@
+package org.allenai.common
+
+import java.util.concurrent.Semaphore
+
+import scala.concurrent._
+import scala.concurrent.duration.Duration
+
+object ParIterator {
+  val queueLimit = 1024
+
+  implicit class ParIteratorEnrichment[T](val input: TraversableOnce[T]) extends AnyVal {
+
+    def parForeach(f: T => Unit)(implicit ec: ExecutionContext): Unit = {
+      // If there are a billion items in the iterator, we don't want to create a billion futures,
+      // so we limit the number of futures we create with this semaphore.
+      val sema = new Semaphore(queueLimit)
+      input.foreach { item =>
+        // Try to pass it off to a future. If no futures are available, do the work
+        // in this thread.
+        val success = sema.tryAcquire()
+        if (success)
+          Future {
+            f(item)
+            sema.release()
```

Comment:
Yes, of course.
On Mar 7, 2015 8:23 AM, "Jesse Kinkead" <notifications@github.com> wrote:

> In core/src/main/scala/org/allenai/common/ParIterator.scala
> <https://github.com/allenai/common/pull/103#discussion_r25997186>:
>
> > +  val queueLimit = 1024
> > +
> > +  implicit class ParIteratorEnrichment[T](val input: TraversableOnce[T]) extends AnyVal {
> > +
> > +    def parForeach(f: T => Unit)(implicit ec: ExecutionContext): Unit = {
> > +      // If there are a billion items in the iterator, we don't want to create a billion futures,
> > +      // so we limit the number of futures we create with this semaphore.
> > +      val sema = new Semaphore(queueLimit)
> > +      input.foreach { item =>
> > +        // Try to pass it off to a future. If no futures are available, do the work
> > +        // in this thread.
> > +        val success = sema.tryAcquire()
> > +        if (success)
> > +          Future {
> > +            f(item)
> > +            sema.release()
>
> Should be in a finally, yes?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/allenai/common/pull/103/files#r25997186>.
>